### PR TITLE
Cortx-29601: RBF for Object tagging versioning

### PR DESCRIPTION
**Pull request by AshwiniS-Borse**
# Problem Statement
- RBF for object tagging versioning test cases.

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [x] Attach test execution logs
-  [x] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by AshwiniS-Borse** [comment]
pytest collection is green.

Execution logs are present in below TE:
https://jts.seagate.com/browse/TEST-43767.
Below 2 tests are failed because of existing sev-3 defects
[TEST-40438](https://jts.seagate.com/browse/TEST-40438)
[TEST-40439](https://jts.seagate.com/browse/TEST-40439)

**Comment by AshwiniS-Borse** [comment]
@apoorva-rao @VikasKumarC @DhobaleVishal @archanabpatil Please review PR

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 1]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
-        get_tag1 = resp[1]
-        for pair in put_tag:
-            assert_utils.assert_in(pair, get_tag1, "Mismatch in tag Key-Value pair."
-                                                  f"Expected: {put_tag} \n Actual: {get_tag1}")
-        LOGGER.info("Step 6: Performed GET Object Tagging for %s with versionId=%s is %s",
-                    self.object_name, latest_ver, get_tag)
+            LOGGER.info("Step 3: Perform PUT Object Tagging for %s with tag set as %s with tag key "
+                        "containing allowed special characters", self.object_name, tag_or)
+            resp = s3_cmn_lib.put_object_tagging(s3_tag_test_obj=self.s3_tag_obj,
+                                                s3_ver_test_obj=self.s3_ver_obj,
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:487]: Let's correct the indentation for 487 - 493

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 2]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
+                                                tag_overrides=tag_or,
+                                                version_id=latest_ver)
+            assert_utils.assert_true(resp[0], resp)
+            put_tag = self.ver_tag[self.object_name][latest_ver]
+            LOGGER.info("Step 3: Performed PUT Object Tagging for %s with tag set as %s with tag "
+                        "key containing allowed special characters", self.object_name, tag_or)
+            LOGGER.info("Step 4: Perform GET Object Tagging for %s with versionId=%s",
+                        self.object_name, latest_ver)
+            resp = s3_cmn_lib.get_object_tagging(s3_tag_test_obj=self.s3_tag_obj,
+                                                s3_ver_test_obj=self.s3_ver_obj,
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:501]: Let's correct the indentation here

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 3]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
+                                                bucket_name=self.bucket_name,
+                                                object_name=self.object_name, version_id=latest_ver)
+            assert_utils.assert_true(resp[0], resp)
+            get_tag = resp[1]
+            for pair in put_tag:
+                assert_utils.assert_in(pair, get_tag, "Mismatch in tag Key-Value pair."
+                                                    f"Expected: {put_tag} \n Actual: {get_tag}")
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:508]: Let's correct indentation here

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 4]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
+            LOGGER.info("Step 4: Performed GET Object Tagging for %s with versionId=%s is %s",
+                        self.object_name, latest_ver, get_tag)
+        #Since all sprcial characters are allowed in RGW, following part of code is invalid
+        if S3_ENGINE_RGW != CMN_CFG["s3_engine"]:
+            tag_or = [{"Key": "tag*key", "Value": "tag1value"},
+                    {"Key": "tag,key", "Value": "tag1value"}]
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:514]: indentation correction needed

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 5]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
+            LOGGER.info("Step 5: Perform PUT Object Tagging for %s with tag set as %s with tag "
+                        "key containing other than allowed special characters", self.object_name,
+                        tag_or)
+            resp = s3_cmn_lib.put_object_tagging(s3_tag_test_obj=self.s3_tag_obj,
+                                                s3_ver_test_obj=self.s3_ver_obj,
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:519]: indentation correction needed

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 6]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
+                                                version_id=latest_ver)
+            assert_utils.assert_false(resp[0], resp)
+            assert_utils.assert_in(err_msg.S3_RGW_BKT_INVALID_TAG_ERR, resp[1].message)
+            LOGGER.info("Step 5: PUT Object Tagging for %s with tag set as %s with tag key "
+                        "containing other than allowed special characters failed as expected",
+                        self.object_name, tag_or)
+            LOGGER.info("Step 6: Perform GET Object Tagging for %s with versionId=%s",
+                        self.object_name, latest_ver)
+            resp = s3_cmn_lib.get_object_tagging(s3_tag_test_obj=self.s3_tag_obj,
+                                                s3_ver_test_obj=self.s3_ver_obj,
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:534]: Let's correct indentation

File: tests/s3/versioning/test_limits_tagging_versioning.py [hunk 7]
```diff
@@ -471,61 +473,72 @@ def test_tag_key_spl_char_41283(self, versioning_status):
...
+                                                bucket_name=self.bucket_name,
+                                                object_name=self.object_name, version_id=latest_ver)
+            assert_utils.assert_true(resp[0], resp)
+            get_tag1 = resp[1]
+            for pair in put_tag:
+                assert_utils.assert_in(pair, get_tag1, "Mismatch in tag Key-Value pair."
+                                                    f"Expected: {put_tag} \n Actual: {get_tag1}")
```
Comment (apoorva-rao) [inline tests/s3/versioning/test_limits_tagging_versioning.py:541]: Let's correct the indentation here

**Review by apoorva-rao — CHANGES_REQUESTED** [review]
@AshwiniS-Borse changes look good, added minor comments inline for indentation

**Review by apoorva-rao — APPROVED** [review]

File: config/s3/s3_config.yaml [hunk 8]
```diff
@@ -75,4 +75,5 @@ delay:
   set_obj_tag: 15
   del_obj_tag: 20
 
-object_tagging_special_char: ["+", "-", "=", ".", "_", ":", "/", "@"]
+object_tagging_special_char_cortx: ["+", "-", "=", ".", "_", ":", "/", "@"]
+object_tagging_special_char_rgw: ["~", "`", "!", "@", "#", "$", "%", "^", "&", "*", "(", ")", "-", "_", "+", "=", ";", ":", "|", "\\", ":", ";", "\"", "'", "<", ",", ">", ".", "?", "/"]
```
Comment (archanabpatil) [inline config/s3/s3_config.yaml:79]: is this mentioned somewhere?

**Review by archanabpatil — COMMENTED** [review]
pls fix codacy

**Review by DhobaleVishal — APPROVED** [review]
@AshwiniS-Borse Make sure changes verified by executing on latest cortx-rgw build..

**Review by VikasKumarC — APPROVED** [review]

**Review by ravindrachoudhari — APPROVED** [review]

**Pull request merged**